go/vt/callerid/callerid.go

```diff
@@ -26,14 +26,15 @@ import (
 )
 
 // The datatype for CallerID Context Keys
-type callerIDKey int
-
-var (
-	// internal Context key for immediate CallerID
-	immediateCallerIDKey callerIDKey
-	// internal Context key for effective CallerID
-	effectiveCallerIDKey callerIDKey = 1
-)
+type callerIDContextKey struct{}
```

Comment:
Previously the keys were integers. Now we have only one key and it is an empty struct. Is it actually more efficient to use an empty struct as the key versus an integer value of 0?

Reply:
An int is 4 or 8 bytes of storage whereas an empty struct is 0 bytes AFAIUI.

Reply:
Yeah, a zero struct occupies no extra space, it's relatively insignificant here since we're talking about a single integer, but a zero sized struct has a constant pointer in the Go runtime and is effectively a pointer to one address no matter the type. Specifically this `runtime.zerobase`.

The comparison has to be done with an interface because of the `any` in `Context`, so an interface just holds a reference to what type + the value, in which case, it's effectively only a ref to the type without a value, since the value is nothing, and the comparison is just on the type, rather than an interface of `{type: int, val: 1}` or whatever.

The docs also suggest using empty structs for keys: https://pkg.go.dev/context#WithValue

> To avoid allocating when assigning to an interface{}, context keys often have concrete type struct{}. Alternatively, exported context key variables' static type should be a pointer or interface. 

Examples of this can be seen through the stdlib codebase.

https://github.com/golang/go/blob/a71bb570d064f6d0c1cf59cef4b7a044be7a39d3/src/runtime/trace/annotation.go#L14
https://github.com/golang/go/blob/a71bb570d064f6d0c1cf59cef4b7a044be7a39d3/src/net/http/transport.go#L2809
https://github.com/golang/go/blob/a71bb570d064f6d0c1cf59cef4b7a044be7a39d3/src/runtime/pprof/label.go#L25

etc

I doubt in practice this makes a tangible difference, but it is the established pattern in stdlib for context keys. If they were to make compiler optimizations for this, I imagine they'd target this pattern specifically.

Reply:
Nice. TiL